File: kluctl/utils/inclusion.py


class Inclusion:
    def __init__(self):
        self.includes = set()
        self.excludes = set()

    def add_include(self, type, value):
        self.includes.add((type, value))

    def add_exclude(self, type, value):
        self.excludes.add((type, value))

    def has_type(self, type):
        for t, v in self.includes:
            if t == type:
                return True
        for t, v in self.excludes:
            if t == type:
                return True
        return False

    def _check_list(self, type_and_values, l):
        for t, v in type_and_values:
            if (t, v) in l:
                return True
        return False

    def check_included(self, type_and_values, exclude_if_not_included=False):
        if not self.includes and not self.excludes:
            return True

        is_included = self._check_list(type_and_values, self.includes)
        is_excluded = self._check_list(type_and_values, self.excludes)

        if exclude_if_not_included:
            if not is_included:
                return False

        if is_excluded:
            return False

        return not self.includes or is_included

File: pkg/utils/inclusion.go
package utils

type InclusionEntry struct {
	Type  string
	Value string
}

type Inclusion struct {
	includes map[InclusionEntry]bool
	excludes map[InclusionEntry]bool
}

func NewInclusion() *Inclusion {
	return &Inclusion{
		includes: map[InclusionEntry]bool{},
		excludes: map[InclusionEntry]bool{},
	}
}

func (inc *Inclusion) AddInclude(typ string, value string) {
	inc.includes[InclusionEntry{typ, value}] = true
}

func (inc *Inclusion) AddExclude(typ string, value string) {
	inc.excludes[InclusionEntry{typ, value}] = true
}

func (inc *Inclusion) HasType(typ string) bool {
	if inc == nil {
		return false
	}
	for e, _ := range inc.includes {
		if e.Type == typ {
			return true
		}
	}
	for e, _ := range inc.excludes {
		if e.Type == typ {
			return true
		}
	}
	return false
}

func (inc *Inclusion) checkList(l []InclusionEntry, m map[InclusionEntry]bool) bool {
	for _, e := range l {
		if _, ok := m[e]; ok {
			return true
		}
	}
	return false
}

func (inc *Inclusion) CheckIncluded(l []InclusionEntry, excludeIfNotIncluded bool) bool {
	if inc == nil {
		return true
	}
	if len(inc.includes) == 0 && len(inc.excludes) == 0 {
		return true
	}

	isIncluded := inc.checkList(l, inc.includes)
	isExcluded := inc.checkList(l, inc.excludes)

	if excludeIfNotIncluded {
		if !isIncluded {
			return false
		}
	}
	if isExcluded {
		return false
	}
	return len(inc.includes) == 0 || isIncluded
}
